[Flang][MLIR][OpenMP] Create a deferred declare target marking process for Bridge.cpp (#78502)
llvm/llvm-project: flang/docs/OpenMP-declare-target.md

# Declare Target OpenMP Dialect To LLVM-IR Lowering

The OpenMP dialect lowering of `declare target` is a little unique currently, as it's not an `operation` and is an `attribute` we process it utilising the LLVM Target lowerings `amendOperation`, which occurs immediately after an operation has been lowered to LLVM-IR. As it can be applicable to multiple different operations, we must
specialise this function for each operation type that we may encounter. Currently this is `GlobalOp`'s and 
`FuncOp`'s.

In the case where we encounter a `FuncOp` our processing is fairly simple, if we're processing the device code, we will finish up our removal of `host` marked functions, anything we could not remove previously we now remove, e.g. if it had a `target` directive in it (which we need to keep a hold of to this point, to actually outline the `target` kernel for device). This hopefully leaves us with only `any`, `device` or undeterminable functions left in the module to lower further, reducing the possibiltiy of device incompatible code being in the module.

> For `GlobalOp`'s, the processing is a little more complex, we currently leverage two OMPIRBuilder functions which we have inherited from Clang and moved to the `OMPIRBuilder` to share across the two compiler frontends `registerTargetGlobalVariable` and `getAddrOfDeclareTargetVar`. These two functions are actually recursive and invoke each other depending on the clauses and options provided to the `OMPIRBuilder` (in particular unified shared memory), but the main functionality they provide is the generation of a new global pointer for device with a "ref_" prefix, and enqueuing metadata generation by the `OMPIRBuilder` at the end of the module, for both host and device that links the newly generated device global pointer and the host pointer together across the two modules (and resulting binaries). 

Reviewer: ```suggestion
For `GlobalOp`'s, the processing is a little more complex. We currently leverage the `registerTargetGlobalVariable` and `getAddrOfDeclareTargetVar` `OMPIRBuilder` functions shared with Clang. These two functions invoke each other depending on the clauses and options provided to the `OMPIRBuilder` (in particular, unified shared memory). Their main purposes are the generation of a new global device pointer with a "ref_" prefix on the device and enqueuing metadata generation by the `OMPIRBuilder` to be produced at module finalization time. This is done for both host and device and it links the newly generated device global pointer and the host pointer together across the two modules. 
```

> Two things of note about the `GlobalOp` processing, the first is that similarly to other metadata (e.g. for `TargetOp`) that is shared across both host and device modules, the device needs access to the previously generated host IR file, which is done through another `attribute` applied to the `ModuleOp` by the compiler frontend. The file is loaded in and consumed by the `OMPIRBuilder` to populate it's `OffloadInfoManager` data structures, keeping host and device appropriately synchronised.

Reviewer: ```suggestion
Similarly to other metadata (e.g. for `TargetOp`) that is shared across both host and device modules, processing of `GlobalOp`'s in the device needs access to the previously generated host IR file, which is done through another `attribute` applied to the `ModuleOp` by the compiler frontend. The file is loaded in and consumed by the `OMPIRBuilder` to populate it's `OffloadInfoManager` data structures, keeping host and device appropriately synchronised.
```

# Current Support

For the moment, `declare target` should work for:
* Marking functions/subroutines and function/subroutine interfaces for generation on host, device or both.
* Implicit function/subroutine capture for calls emitted in a `target` region or explicitly marked `declare 
   target` function/subroutine. Note: Calls made via arguments passed to other functions must still be 
   themselves marked `declare target`, e.g. passing a `C` function pointer and invoking it, then the interface
   and the `C` function in the other module must be marked `declare target`, with the same type of 
   marking as indicated by the specification.
* Marking global variables with `declare target`'s `link` clause and mapping the data to the device data 

>    environment utilising `declare target` (may not work for all types yet, but for scalars and arrays

Reviewer: ```suggestion
   environment utilising `declare target`. This may not work for all types yet, but for scalars and arrays
```

>    of scalars, it should).

Reviewer: ```suggestion
   of scalars, it should.
```

Doesn't work for, or needs further testing for:
* Marking the following types with `declare target link` (needs further testing):
    * Descriptor based types, e.g. pointers/allocatables.
    * Derived types.
    * Members of derived types (use-case needs legality checking with OpenMP specification).

> * Marking global variables with `declare target`'s `to` clause, a lot of the lowering should exist, but it needs further testing and likely some further changes to fully function.

Reviewer: ```suggestion
* Marking global variables with `declare target`'s `to` clause. A lot of the lowering should exist, but it needs further testing and likely some further changes to fully function.
```

as `declare target` and then map it, embedding it into the device IR and 
assigning to it. 

```Fortran
function func_t_device() result(i)
    !$omp declare target to(func_t_device) device_type(nohost)
        INTEGER :: I
        I = 1
end function func_t_device

program main
!$omp target
    call func_t_device()
!$omp end target
end program
```

In the above example, we are stating that a function is required on device
utilising `declare target`, and that we will not be utilising it on host, 
so we are in theory free to remove or ignore it there. A user could also 
in this case, leave off the `declare target` from the function and it 
would be implicitly marked `declare target any` (for both host and device), 
as it's been utilised within a target region.

# Declare Target as represented in the OpenMP Dialect

In the OpenMP Dialect `declare target` is not represented by a specific 
`operation`. Instead it's a OpenMP dialect specific `attribute` that can be 
applied to any operation in any dialect. This helps to simplify the 
utilisation of it, instead of replacing or modifying existing global or 
function `operations` in a dialect it applies to it as extra metadata that 

> the lowering can use in different ways as is necesary. 

Reviewer: ```suggestion
the lowering can use in different ways as is necessary. 
```

declareTargetGlobal.isDeclareTarget();
```

# Declare Target Fortran OpenMP Lowering

The initial lowering of `declare target` to MLIR for both use-cases is done
inside of the usual OpenMP lowering in flang/lib/Lower/OpenMP.cpp. However some
direct calls to `declare target` related functions from Flang's lowering bridge 
in flang/lib/Lower/Bridge.cpp are made.

The marking of operations with the declare target attribute happens in two 
phases, the second one optional and contingent on the first failing. The 
initial phase happens when the declare target directive and its clauses 
are initially processed, with the primary data gathering for the directive and
clause happening in a function called `getDeclareTargetInfo`. This is then used
to feed the `markDeclareTarget` function, which does the actual marking 
utilising the `DeclareTargetInterface`. If it encounters a variable or function
that has been marked twice over multiple directives with two differing device
types (e.g. `host`, `nohost`), then it will swap the device type to `any`.

Whenever we invoke `genFIR` on an `OpenMPDeclarativeConstruct` from the 
lowering bridge, we are also invoking another function called 

> `gatherOpenMPDeferredDeclareTargets` which gathers information relevant to the

Reviewer: ```suggestion
`gatherOpenMPDeferredDeclareTargets`, which gathers information relevant to the
```

application of the `declare target` attribute. This information 
includes the symbol that it should be applied to, device type clause, 
and capture clause, and it is stored in a vector that is part of the lowering
bridge's instantiation of the `AbstractConverter`. It is only stored if we 
encounter a function or variable symbol that does not have an operation 
instantiated for it yet. This cannot happen as part of the 
initial marking as we must store this data in the lowering bridge and we 
only have access to the abstract version of the converter via the OpenMP 
lowering. 

The information produced by the first phase is used in the second phase, 
which is a form of deferred processing of the `declare target` marked 
operations that have delayed generation and cannot be proccessed in the 
first phase. The main notable case this occurs currently is when a 
Fortran function interface has been marked. This is 
done via the function 

> `markOpenMPDeferredDeclareTargetFunctions` which is called from the lowering

Reviewer: ```suggestion
`markOpenMPDeferredDeclareTargetFunctions`, which is called from the lowering
```

through another `attribute` applied to the `ModuleOp` by the compiler 
frontend. The file is loaded in and consumed by the `OMPIRBuilder` to 
populate it's `OffloadInfoManager` data structures, keeping host and 
device appropriately synchronised.

The second (and more important to remember) is that as we effectively replace
the original LLVM-IR generated for the `declare target` marked `GlobalOp` we
have some corrections we need to do for `TargetOp`'s (or other region 
operations that use them directly) which still refer to the original lowered
global operation. This is done via `handleDeclareTargetMapVar` which is invoked
as the final function and alteration to the lowered `target` region, it's only
invoked for device as it's only required in the case where we have emitted the
"ref" pointer , and it effectively replaces all uses of the originally lowered
global symbol, with our new global ref pointer's symbol. Currently we do not
remove or delete the old symbol, this is due to the fact that the same symbol
can be utilised across multiple target regions, if we remove it, we risk 
breaking lowerings of target regions that will be processed at a later time. 

> To appropriately delete these no longer necesary symbols we would need a 

Reviewer: ```suggestion
To appropriately delete these no longer necessary symbols we would need a 
```
